src/qiskit_qasm3_import/state.py

```diff
@@ -0,0 +1,205 @@
+        """Set the addressing mode to virtual.
+        On success return `True`, otherwise raise an exception."""
+        if self._state == self.PHYSICAL:
+            raise_from_node(
+                node,
+                "Virtual qubit declared in physical addressing mode. Mixing modes not currently "
+                "supported.",
+            )
+        self._state = self.VIRTUAL
+
+
+class SymbolTable:
+    __slots__ = ("symbols", "scope", "base", "global_symbols")
+
+    def __init__(self, scope: Scope, base: Optional["SymbolTable"] = None):
+        self.symbols = {}
+        self.scope = scope
+        self.base = base
+        self.global_symbols = base.global_symbols if base is not None else _BUILTINS.copy()
+
+    def __contains__(self, name: str):
+        return name in (self.symbols, self.global_symbols) or (
+            self.base is not None and name in self.base.symbols
+        )
+
+    def get(self, name: str, node=None):
+        """Lookup symbol `name`
+
+        Return `None` if no symbol is found. Raise an exception if the symbol is
+        inaccessible in the scope of this table.
+        """
+        if (symbol := self.symbols.get(name, None)) is not None:
+            return symbol
+        if self.base is not None:
+            if (symbol := self.base.get(name, None)) is None:
+                return None
+            return self._check_visible(symbol, node)
+        return self.global_symbols.get(name, None)
+
+    def _check_visible(self, symbol, node):
+        if (
+            symbol.scope is Scope.BUILTIN
+            or self.scope is not Scope.GATE
+            or symbol.scope is not Scope.GLOBAL
+        ):
+            return symbol
+        if self.scope is Scope.GATE:
+            if (
+                isinstance(symbol.type, types.Gate)
+                or isinstance(
+                    symbol.type, (types.Int, types.Uint, types.Float, types.Angle, types.Duration)
+                )
+                and symbol.type.const
+            ):
+                return symbol
+        if isinstance(symbol.type, types.HardwareQubit):
+            raise PhysicalQubitInGateError(node.name, node)
+        raise_from_node(node, f"Symbol {symbol.name} is not visible in the scope of a gate")
+
+    def insert(self, symbol):
+        target = self.global_symbols if symbol.type == types.HardwareQubit() else self.symbols
+        target[symbol.name] = symbol
+
+    def globals(self):
+        if self.global_symbols is None:
+            return []
+        return self.global_symbols.values()
+
+
+class State:
+    __slots__ = ("scope", "source", "circuit", "symbol_table", "_unique", "addressing_mode")
+
+    def __init__(self, source: Optional[str] = None):
+        # We use the entire source, because at the moment, that's what all the error messages
+        # expect; the nodes have references to the complete source in their spans.
+        self.source = source
+        self.init_inner()
+
+    def init_inner(
```

Comment:
Super minor, but probably let's make this private (`_init_inner`).  I don't think I was especially strict on that split when I wrote most of the package, but this one feels fairly cut-and-dried.

This DRY form looks much cleaner that what I'd originally written though, so thanks for that!

Reply:
That's what I had originally, but removing the underscore was the fastest way to silence a spurious error: `W0212: Access to a protected member _init_inner of a client class`. Better to suppress the linter here.

Reply:
Fixed in 33ee260

Reply:
Oh yeah, I see.  I agree the lint suppression is the right call there - it's not really a client, it's a part of the constructor method.